### DISCUSSION TITLE
Running Conky, from root but by using 'su' to invoke an unprivileged user, produces strange results

### DISCUSSION THREAD

**Discussion by LinuxOnTheDesktop** [brndnmtthws/conky#1365] [Q&A]
Hi there.

I have a Bash script, which runs as root (for various reasons) and which then runs a _second_ Bash script as an unprivileged user. The second script runs Conky. (Bear with me!) When I do this, Conky does run, but various (yet further!) helper programs (these ones written in Lua, and called from Conky) fail to print to the screen. When I run Conky _normally_, those helper programs work fine. Here is how the aforementioned first Bash script runs Conky:

```
userMain=$(grep 1000 /etc/passwd | sed "s/:.*$//g") ||
    msg_error_fatal "failed to identify main user"
su -c ${conkyRestartScript} "${userMain}"
```

Do I need to add some other option to `su`? I would appreciate any help.

**Comment by LinuxOnTheDesktop** [comment]
I discover - by setting Conky to print errors - that various necessary environmental variables are not being set. Those variables comprise or at least include: `LUA_PATH`; `DBUS_SESSION_BUS_ADDRESS`; `XDG_RUNTIME_DIR`.

**Comment by LinuxOnTheDesktop** [comment]
I have not really solved the problem, but I have a workaround. To wit: no longer the script that I was running privileged as privileged but instead as a normal user. Doing that still allows the script to work - it is just that it prompts several times (once graphically, once in a terminal) for elevation. And now I can (and do) restart Conky as a normal user, and everything works.

**Comment by Caellian** [comment]
**WARNING:** If you use conky as su, I'd warn against using online scripts without **thoroughly** checking them as Lua can execute arbitrary code and mess up your data/installation.

Hmm, `XDG_RUNTIME_DIR` is usually set by the WM and/or DM. You can set it to /root (or whatever the home of your root user is), `DBUS_SESSION_BUS_ADDRESS` can be set by running `export $(dbus-launch)`, as for `LUA_PATH` 🤷🏻‍♂️, I don't have it set but I'm guessing it's a path to lua libraries (maybe not needed?).

I think writing a launch script for conky that sets those up like:
```sh
export XDG_RUNTIME_DIR=/root
export $(dbus-launch)
# ...
conky
```
is better than having to enter your password for everything. But also, check into why your WM isn't setting those up for you. What are you using?

**Comment by LinuxOnTheDesktop** [comment]
I aim to investigate the matter soon (when I have time; though I think I may have abandoned the idea at issue). In the meantime: thanks; I'm using Linux Mint Cinnamon (version 21), and do note that the proposal is to use `su` to run Conky as a particular user, not as root.

**Comment by Caellian** [comment]
Hmm, I was wrong about the `XDG_RUNTIME_DIR`. Found a stack overflow question that explains it better: https://stackoverflow.com/questions/28809235/how-to-get-xdg-variables-with-sudo
